arch-design/api/UnifiedAPI.md

```diff
@@ -0,0 +1,208 @@
+
+
+##### Non Functional Requirements
+
+1. The system should be highly available. Multiple users can login to the system and provision resources like, create /volume/fileshare/buckets etc.
+
+2. The system should work with minimum latency.
+
+3. The system should be scalable and it should handle the high amount of users' requests.
+
+## Proposal
+
+To support Unified API for SODA projects, all other project's API can be added to one project named API-Server. Currently API-Server has Hotpot(on-prem storage) related APIs, now other projects like multi-cloud API can also be added to API-Server project which inherently will become similar to API gateway for SODA projects. Further all the request calls coming to API-Server can be validated and forwarded to the Controller project.
+
+Controller is a project, where all the intelligence logics related to selection of on-prem backend storages are there. Basically, this selection logic helps to pick one storage among multiple registered storages based on profile. Later that selected storage can be used for provisioning etc. Now the selection logic can be enhanced to select one storage for provisioning among on-prem and cloud storages. Here selection logic can take input from profile and tiers.
+
+Once the backend storage is selected, calls can be forwarded to either Dock or Multi-cloud API for further processing.
+
+So the overall workflow should looks like:
+
+a. Clients <---> API-Server <---> Controller(<---> DB) ---> Dock <---> Storage Backend
+
+b. Clients <---> API-Server <---> Controller <---> Multi-Cloud API <---> S3-service(<--->DB) <---> Cloud backend
+
+c. Clients <---> API-Server <---> Controller ---> Multi-Cloud API <---> Backend-service(<--->DB) <---> on-prem/cloud backend
+
+In above workflow, two points need to note:
+  * For GET API call to hotpot, the results are coming from DB
+
+  * For GET API call to Multi-cloud, the results are coming from Multi-cloud API(i.e, in this case there is no direct access of DB from Controller, Controller has to make Call to Multi-cloud API only for any request)  
+
+
+The below [sequence diagrams](#Sequence) can be referred to see the end-to-end flow in more detail.
+
+Please find the details of proposed diagram
+#### Fig. I: The proposed architecture diagram
+![](resources/unified-3.png)
+
+### Sequence Diagrams:
+1. POST http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+  ![](resources/seq-1.png)
+
+2. POST http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+  ![](resources/seq-2.png)
+
+3. POST http://HOST_IP:PORT/v1beta/{tenantId}/block/xxx
+
+   POST http://HOST_IP:PORT/v1beta/{tenantId}/file/xxx
+
+  ![](resources/seq-3.png)
+
+4. GET http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+   GET http://HOST_IP:PORT/v1beta/{tenantId}/{block or file}/xxx
+
+  ![](resources/Get-se-1.png)
+
+
+### Data model impact
+
+New data model will be added to the API server and Controller module. Here are some examples:
+
+```
+type Backend struct {
```

Comment:
Why duplicate these models? Are we going to store all Bucket and Backend information at two places?